codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2DataFrame.java

```diff
@@ -0,0 +1,173 @@
+    }
+
+    /**
+     * Construct a new data message.
+     *
+     * @param content non-{@code null} payload
+     * @param endStream whether this data should terminate the stream
+     * @param padding additional bytes that should be added to obscure the true content size
+     */
+    public DefaultHttp2DataFrame(ByteBuf content, boolean endStream, int padding) {
+        this.content = checkNotNull(content, "content");
+        this.endStream = endStream;
+        if (padding < 0 || padding > Http2CodecUtil.MAX_UNSIGNED_BYTE) {
+            throw new IllegalArgumentException("padding must be non-negative and less than 256");
+        }
+        this.padding = padding;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame setStream(Object stream) {
+      super.setStream(stream);
+      return this;
+    }
+
+    @Override
+    public boolean endStream() {
+        return endStream;
+    }
+
+    @Override
+    public int padding() {
+        return padding;
+    }
+
+    @Override
+    public ByteBuf content() {
+        if (content.refCnt() <= 0) {
+            throw new IllegalReferenceCountException(content.refCnt());
+        }
+        return content;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame copy() {
+        return new DefaultHttp2DataFrame(content().copy(), endStream, padding);
+    }
+
+    @Override
+    public DefaultHttp2DataFrame duplicate() {
+        return new DefaultHttp2DataFrame(content().duplicate(), endStream, padding);
+    }
+
+    @Override
+    public int refCnt() {
+        return content.refCnt();
+    }
+
+    @Override
+    public boolean release() {
+        return content.release();
+    }
+
+    @Override
+    public boolean release(int decrement) {
+        return content.release(decrement);
+    }
+
+    @Override
+    public DefaultHttp2DataFrame retain() {
+        content.retain();
+        return this;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame retain(int increment) {
+        content.retain(increment);
+        return this;
+    }
+
+    @Override
+    public String toString() {
+        return "DefaultHttp2DataFrame(stream=" + stream() + ",content=" + content
+            + ",endStream=" + endStream + ",padding=" + padding + ")";
+    }
+
+    @Override
+    public DefaultHttp2DataFrame touch() {
+        content.touch();
+        return this;
+    }
+
+    @Override
+    public DefaultHttp2DataFrame touch(Object hint) {
+        content.touch(hint);
+        return this;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (!(o instanceof DefaultHttp2DataFrame)) {
+            return false;
+        }
+        DefaultHttp2DataFrame other = (DefaultHttp2DataFrame) o;
+        return super.equals(other) && content.equals(other.content())
+            && endStream == other.endStream && padding == other.padding;
+    }
+
+    @Override
+    public int hashCode() {
+        int hash = super.hashCode();
+        hash = hash * 31 + content.hashCode();
```

Comment:
If it is released, how could I implement equals? I'd argue that using it after the `ByteBuf` is released would be invalid behavior because that means you should have increased the reference count (by definition) and because there is no reasonable implementation. If it needs to continue working after released, then `hashCode()` and `equals()` could probably check to see if the bytebuf was released and then always return `0` and `false`, respectively.

The other option is these classes don't implement `hashCode` and `equals()`. It is there just for testing, but that is useful.

I see that `DefaultFullHttpRequest` detects released buffers for `hashCode()` but not `equals()`. That's not good enough for me, given that it will randomly break (when there is a hash code collision).